File: CPP/ByTier/D/1904_D2_Set_To_Max_Hard_Version.cpp
// Problem: CF 1904 D2 - Set To Max (Hard Version)
// https://codeforces.com/contest/1904/problem/D2

/*
Algorithm:
This problem uses a deque-based approach to simulate the operations and
determine if array `a` can be transformed into array `b`. The key insight is
that we can only set a segment to the maximum value in that segment. So we
process the array from left to right and right to left, using a deque to
maintain the maximum value in the current window and check if the current
position can be set to the target value.

Time Complexity: O(n)
Space Complexity: O(n)

Methods:
- Two passes through the array (left to right and right to left)
- Using a deque to maintain the maximum element in the current window
- For each position, we check if we can set it to the desired value in `b`
*/
#include <algorithm>
#include <deque>
#include <iostream>
#include <stdio.h>
#include <string>
#include <type_traits>
#include <vector>

using namespace std;
namespace Fread {
const int SIZE = 1 << 16;
char buf[SIZE], *S, *T;
inline char getchar() {
  if (S == T) {
    T = (S = buf) + fread(buf, 1, SIZE, stdin);
    if (S == T)
      return '\n';
  }
  return *S++;
}
} // namespace Fread

void solve();
int main() {
  int t = 1;
  cin >> t;
  while (t--)
    solve();
}
void solve() {
  int n;
  cin >> n;
  vector<int> a(n + 2), b(n + 2);
  for (int x = 1; x <= n; x++)
    cin >> a[x];
  for (int x = 1; x <= n; x++)
    cin >> b[x];
  deque<int> dq;
  for (int x = 1; x <= n; x++) {
    // Remove elements from back that are smaller than or equal to current a[x]
    while (!dq.empty() && dq.back() <= a[x])
      dq.pop_back();
    // Remove elements from front that are greater than current b[x]
    while (!dq.empty() && dq.front() > b[x])
      dq.pop_front();
    // Add current a[x] to the back of deque
    dq.push_back(a[x]);
    // If the front of deque equals b[x], we can set a[x] to b[x]
    if (dq.front() == b[x])
      a[x] = b[x];
  }
  // Process from right to left
  for (int x = n; x >= 1; x--) {
    // Same logic as above
    while (!dq.empty() && dq.back() <= a[x])
      dq.pop_back();
    while (!dq.empty() && dq.front() > b[x])
      dq.pop_front();
    dq.push_back(a[x]);
    if (dq.front() == b[x])
      a[x] = b[x];
  }
  // Check if all elements in a match b
  for (int x = 1; x <= n; x++)
    if (a[x] != b[x]) {
      cout << "NO" << endl;
      return;
    }
  cout << "YES" << endl;
}

// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1904/1904_D2_Set_To_Max_Hard_Version.py
# Problem: CF 1904 D2 - Set To Max (Hard Version)
# https://codeforces.com/contest/1904/problem/D2

"""
Algorithm: 
- This problem uses a sliding window technique with a deque to simulate the operations.
- For each position in the array, we maintain a deque that stores values in a decreasing order.
- We process the array from left to right and then from right to left.
- At each step, we check if we can assign the value from b[i] to a[i] based on the constraints of the operations.

Time Complexity: O(n) - Each element is pushed and popped from the deque at most once.
Space Complexity: O(n) - The deque can store up to n elements.
"""

from sys import stdin

input = lambda: stdin.buffer.readline().decode().strip()
from collections import deque

for _ in range(int(input())):
    n = int(input())
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))

    def solve(order):
        dq = deque()
        for i in order:
            # Remove elements from the end of deque that are <= a[i]
            while dq and dq[-1] <= a[i]:
                dq.pop()
            # Remove elements from the front of deque that are > b[i]
            while dq and dq[0] > b[i]:
                dq.popleft()
            # Add a[i] to the deque
            dq.append(a[i])
            # If the front of deque equals b[i], we can assign it to a[i]
            if dq[0] == b[i]:
                a[i] = b[i]

    # Process from left to right
    solve(range(n))
    # Process from right to left
    solve(reversed(range(n)))
    # Check if arrays are equal
    print("YES" if a == b else "NO")


# https://github.com/VaHiX/CodeForces/